src/components/ivdcm_module/net/src/linux_tun_adapter.cc

```diff
@@ -0,0 +1,251 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "net/linux_tun_adapter.h"
+
+#include <arpa/inet.h>
+#include <cstdlib>
+#include <cstring>
+#include <fcntl.h>
+#include <errno.h>
+#include <linux/if_tun.h>
+#include <net/if.h>
+#include <netinet/in.h>
+#include <string>
+#include <sstream>
+#include <sys/ioctl.h>
+#include <sys/socket.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+#include <unistd.h>
+
+#include "utils/logger.h"
+
+namespace net {
+
+CREATE_LOGGERPTR_GLOBAL(logger_, "IVDCM")
+
+LinuxTunAdapter::LinuxTunAdapter(const std::string& nic)
+    : nic_(nic),
+      fds_() {
+}
+
+bool LinuxTunAdapter::RunCommand(int cmd, ifreq *ifr) const {
+  LOG4CXX_AUTO_TRACE(logger_);
+  int fd = socket(AF_INET, SOCK_DGRAM, 0);
+  if (fd == -1) {
+    std::string error(strerror(errno));
+    LOG4CXX_ERROR(logger_, "Could not create socket with error: " << error);
+    return false;
+  }
+  int ret = ioctl(fd, cmd, ifr);
+  if (ret == -1) {
+    std::string error(strerror(errno));
+    LOG4CXX_ERROR(logger_, "Could not run command with error: " << error);
+  }
+  close(fd);
+  return ret != -1;
+}
+
+std::string LinuxTunAdapter::GetName(int id) const {
+  std::ostringstream name;
+  name << nic_ << id;
+  return name.str();
+}
+
+void LinuxTunAdapter::InitRequest(int id, ifreq *ifr) const {
+  LOG4CXX_AUTO_TRACE(logger_);
+  memset(ifr, 0, sizeof(*ifr));
+  std::string name = GetName(id);
+  strncpy(ifr->ifr_name, name.c_str(), sizeof(ifr->ifr_name));
+}
+
+/**
+ * Converts a string into an Internet address stored in a structure
+ * @param value of  Internet address (support only IPv4)
+ * @param addr pointer to save result
+ */
+static void StringToSockAddr(const std::string& value, sockaddr *addr) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  sockaddr_in *addr_in = reinterpret_cast<sockaddr_in*>(addr);
+  addr_in->sin_family = AF_INET;
+  inet_aton(value.c_str(), &addr_in->sin_addr);
+}
+
+int LinuxTunAdapter::NextId() {
+  static int id = 0;
+  return ++id;
+}
+
+/**
+ * Converts an Internet address into a string
+ * @param addr to save result
+ * @param value of  Internet address (support only IPv4)
+ */
+static void SockAddrToString(const sockaddr *addr, std::string *value) {
+  LOG4CXX_AUTO_TRACE(logger_);
+  const sockaddr_in *sin = reinterpret_cast<const sockaddr_in*>(addr);
+  *value = std::string(inet_ntoa(sin->sin_addr));
+}
+
+int LinuxTunAdapter::Create() {
+  LOG4CXX_AUTO_TRACE(logger_);
+  int fd = open("/dev/net/tun", O_RDWR);
+  if (fd == -1) {
+    std::string error(strerror(errno));
+    LOG4CXX_ERROR(logger_, "Could not open /dev/net/tun with error: " << error);
+    return -1;
+  }
+  ifreq ifr;
+  int id = NextId();
+  InitRequest(id, &ifr);
+  ifr.ifr_flags = IFF_TUN | IFF_NO_PI;
+  int ret = ioctl(fd, TUNSETIFF, &ifr);
```

Comment:
may be it is better to create a separate function like InterfaceUp and InterfaceDown?